tests/integration/acp/add_policy/with_extra_perms_and_relations_test.go

```diff
@@ -25,7 +23,7 @@ func TestACP_AddPolicy_ExtraPermissionsAndExtraRelations_ValidPolicyID(t *testin
 
 		Actions: []any{
 			testUtils.AddPolicy{
-				Identity: immutable.Some(1),
+				Identity: testUtils.UserIdentity(1),
```

Comment:
suggestion: I would like to avoid this if possible.

Reply:
But we want to differenciate between node identities and externally provided udentities. How else do you want to do it?
Yes, this make it more verbose, but explicit at the same time. You like "explicit", don't you? :)

Reply:
> But we want to differenciate between node identities and externally provided identities. How else do you want to do it?

I thought I described my prefer way to do it here already and that you were okay with it? https://github.com/sourcenetwork/defradb/pull/3125#discussion_r1803180817

In short, leave previous identities the way they are and make a new one for node Identities rather than trying to have to both fit the same mold.

Reply:
You mean to add to every test action an additional field? Like:
```go
Identity immutable.Option[int]
NodeIdentity immutable.Option[int]
```
I don't think it's good design. First of all, you can't have both of them set, you will have to document every action and check on executtion of this action if not more then 1 is set. Another point is, what if in the future we add another type of identity? That's going to hurt.

And anyway all identities we used so far ARE node identities, beucase they are attached to a node. So it should be:
```go
Identity immutable.Option[int]
UserIdentity immutable.Option[int]
```

Reply:
That is not what I am proposing.

I am saying leave all actions that existed before to use the identity implementation that existed before and not touch them, all tests should remain the way they were. Just rename the `identities [][]identity.Identity` on `state` struct (i.e. `s.idenities`) to `requestingIdentities` (or `userIdentities`) and add a new variable to the `state` struct called `nodeIdentities ` (or whatever you would like to call it, and make the implementation you have introduced in this PR for example 1D array, do all of that for that identity).

Then only the test.cases that use node identity stuff can just use that new variables identity logic for example: `testUtils.GetNodeIdentity` action can operate over just the `nodeIdentities` and doesn't need to worry about the other identities logic the way we had it before.

This was you don't need the new custom `IdentRef` and know for sure each action case is operating on it's domain of identities just like you said you can't have both of them set at once, so I am very much agains't mixing these.

Reply:
There is a problem with keeping things as is because when we start needing to set node identities to docs and/or policies, things will need to be reworked anyways.

One thing I like about the test changes in this PR is the decoupling of identity generation from token generation. It has the potential to keep things cleaner.

I think the `identRef` approach could be improved though and I have a few ideas to simplify things and make both Shahzad and Islam happy with the outcome. We can discuss during standup.

Reply:
> There is a problem with keeping things as is because when we start needing to set node identities to docs and/or policies, things will need to be reworked anyways.

What do you mean by setting node identity to docs/policy? I thought node identity is an identity that will be set once for a node on start (if an admin user chooses to ofc).


> One thing I like about the test changes in this PR is the decoupling of identity generation from token generation. It has the potential to keep things cleaner.

I like the decoupling as well (specially for the type of identity that needs it).

> I think the `identRef` approach could be improved though and I have a few ideas to simplify things and make both Shahzad and Islam happy with the outcome. We can discuss during standup.

I think a middle approach can be to make an optional type that takes a custom type containing the index and the type of identity (i.e. remove the `hasValue` and value type from the custom type). However I just still fail to quite understand why we have to couple both types of identities together